src/vivarium/framework/results/context.py

```diff
@@ -181,13 +182,21 @@ def gather_results(
             else:
                 if stratifications is None:
                     for observation in observations:
-                        df = observation.results_gatherer(filtered_pop)
-                        yield df, observation.name, observation.results_updater
+                        if not observation.to_observe(event):
+                            yield None, None, None
```

Comment:
This always confuses me. Why do we yield instead of return?

Reply:
computational efficiency. We _could_ combine these datasets into a mega-list and return it all at once and then loop through as needed later. But generators allow for lazy evaluation so we don't need to store all that at once in memory